File: tests/regenerate/gen_normal_distribution.c
/*
 * Generator for phase-2 reference tables: full cdfnor end-to-end.
 *
 * Writes (mean, sd, x, p, q) rows from cdfnor with which=1 over a small
 * grid spanning the body and tails. Verifies that the Rust Normal
 * composition matches the C cdfnor (which is, in turn, the composition
 * of cumnor with the standardization).
 */

#include "refs/cdflib.h"
#include <stdio.h>
#include <stdlib.h>

static FILE* open_csv(const char* relpath, const char* header) {
    FILE* f = fopen(relpath, "w");
    if (!f) {
        fprintf(stderr, "could not open %s for writing\n", relpath);
        exit(1);
    }
    fprintf(f, "# %s\n", header);
    return f;
}

int main(void) {
    /* --------------------------------------------------------- normal cdf */
    FILE* f = open_csv("tests/data/normal_cdf.csv", "mean, sd, x, cdf, sf");
    double means[] = {0.0, -1.5, 3.25};
    double sds[]   = {1.0,  0.5, 2.75};
    for (size_t i = 0; i < sizeof(means)/sizeof(means[0]); i++) {
        double mean = means[i];
        double sd   = sds[i];
        for (double x = mean - 12.0 * sd; x <= mean + 12.0 * sd + 1e-12; x += sd * 0.125) {
            int which = 1;
            int status = 0;
            double p = 0.0, q = 0.0, xx = x, mm = mean, ss = sd, bound = 0.0;
            cdfnor(&which, &p, &q, &xx, &mm, &ss, &status, &bound);
            if (status != 0) continue;
            fprintf(f, "%.17g,%.17g,%.17g,%.17g,%.17g\n", mean, sd, x, p, q);
        }
    }
    fclose(f);

    /* ------------------------------------------------------- normal inv cdf */
    /* Parametrize by x then store (mean, sd, p, q, x). At x = mean + 8·sd,
       q is ~6e-16, unrepresentable as 1-p. */
    f = open_csv("tests/data/normal_inverse_cdf.csv", "mean, sd, p, q, x");
    for (size_t i = 0; i < sizeof(means)/sizeof(means[0]); i++) {
        double mean = means[i];
        double sd   = sds[i];
        for (double x = mean - 7.0 * sd; x <= mean + 7.0 * sd + 1e-12; x += sd * 0.125) {
            int which = 1;
            int status = 0;
            double p = 0.0, q = 0.0, xx = x, mm = mean, ss = sd, bound = 0.0;
            cdfnor(&which, &p, &q, &xx, &mm, &ss, &status, &bound);
            if (status != 0) continue;
            if (p <= 0.0 || p >= 1.0 || q <= 0.0 || q >= 1.0) continue;
            fprintf(f, "%.17g,%.17g,%.17g,%.17g,%.17g\n", mean, sd, p, q, x);
        }
    }
    fclose(f);

    fprintf(stderr, "wrote 2 tables under tests/data/\n");
    return 0;
}

File: tests/regenerate/gen_normal_distribution.f90
! Reference tables for the Normal distribution via cdfnor.
!
! Build and run via tests/regenerate/regenerate.sh.

program gen_normal_distribution
  implicit none
  integer, parameter :: rk = kind(1.0d0)
  external :: cdfnor

  real(kind=rk), dimension(3) :: means = (/ 0.0_rk, -1.5_rk, 3.25_rk /)
  real(kind=rk), dimension(3) :: sds   = (/ 1.0_rk,  0.5_rk, 2.75_rk /)
  integer :: i, unit, which, status
  real(kind=rk) :: x, mean, sd, p, q, bound
  real(kind=rk) :: xx, mm, ss

  ! normal_cdf
  open(newunit=unit, file='tests/data/normal_cdf.csv', status='replace', action='write')
  write(unit, '(a)') '# mean, sd, x, cdf, sf'
  do i = 1, 3
    mean = means(i)
    sd = sds(i)
    x = mean - 12.0_rk * sd
    do while (x <= mean + 12.0_rk * sd + 1.0e-12_rk)
      which = 1; status = 0; p = 0.0_rk; q = 0.0_rk; bound = 0.0_rk
      xx = x; mm = mean; ss = sd
      call cdfnor(which, p, q, xx, mm, ss, status, bound)
      if (status == 0) then
        call putval(unit, mean, .false.)
        call putval(unit, sd, .false.)
        call putval(unit, x, .false.)
        call putval(unit, p, .false.)
        call putval(unit, q, .true.)
      end if
      x = x + sd * 0.125_rk
    end do
  end do
  close(unit)

  ! normal_inverse_cdf: generate by x, store (mean, sd, p, q, x).
  open(newunit=unit, file='tests/data/normal_inverse_cdf.csv', status='replace', action='write')
  write(unit, '(a)') '# mean, sd, p, q, x'
  do i = 1, 3
    mean = means(i)
    sd = sds(i)
    x = mean - 7.0_rk * sd
    do while (x <= mean + 7.0_rk * sd + 1.0e-12_rk)
      which = 1; status = 0; p = 0.0_rk; q = 0.0_rk; bound = 0.0_rk
      xx = x; mm = mean; ss = sd
      call cdfnor(which, p, q, xx, mm, ss, status, bound)
      if (status == 0 .and. p > 0.0_rk .and. p < 1.0_rk &
          .and. q > 0.0_rk .and. q < 1.0_rk) then
        call putval(unit, mean, .false.)
        call putval(unit, sd, .false.)
        call putval(unit, p, .false.)
        call putval(unit, q, .false.)
        call putval(unit, x, .true.)
      end if
      x = x + sd * 0.125_rk
    end do
  end do
  close(unit)

  write(0, '(a)') 'wrote 2 tables under tests/data/'

contains

  subroutine putval(unit, v, last)
    integer, intent(in) :: unit
    real(kind=rk), intent(in) :: v
    logical, intent(in) :: last
    character(len=32) :: buf
    write(buf, '(es25.17e3)') v
    if (last) then
      write(unit, '(a)') trim(adjustl(buf))
    else
      write(unit, '(a, a)', advance='no') trim(adjustl(buf)), ','
    end if
  end subroutine putval

end program gen_normal_distribution
